src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Summarize.cs

```diff
@@ -0,0 +1,183 @@
+                {
+                    args[i] = TexlStrings.SummarizeArg2;
+                }
+                else
+                {
+                    args[i] = TexlStrings.SummarizeArg3;
+                }
+
+                addSwitch = !addSwitch;
+            }
+
+            return new[] { args };
+        }
+
+        public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType returnType, out Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+        {
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+            Contracts.Assert(MinArity <= args.Length && args.Length <= MaxArity);
+
+            bool isValid = base.CheckTypes(context, args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+            Contracts.Assert(returnType.IsTable);
+
+            // Source table type
+            DType sourceType = argTypes[0];
+
+            if (!sourceType.IsTable)
+            {
+                sourceType = DType.EmptyTable;
+            }
+
+            var atLeastOneGroupByColumn = false;
+
+            for (int i = 1; i < args.Length; i++)
+            {
+                var argType = argTypes[i];
+                var arg = args[i];
+
+                DName columnName;
+                DType existingType;
+
+                // All args starting at index 1 need to be identifiers or aggregate functions (wrapped in AsNode node).
+                switch (arg)
+                {
+                    case AsNode nameNode:
+                        existingType = argType;
+                        columnName = nameNode.Right.Name;
+                        break;
+
+                    case FirstNameNode:
+                        if (!TryGetColumnLogicalName(sourceType, context.Features.SupportColumnNamesAsIdentifiers, arg, errors, out columnName, out existingType))
+                        {
+                            isValid = false;
+                            continue;
+                        }
+
+                        atLeastOneGroupByColumn = true;
+                        break;
+
+                    default:
+                        isValid = false;
+                        errors.EnsureError(DocumentErrorSeverity.Severe, arg, TexlStrings.ErrNotSupportedFormat_Func, Name);
+                        continue;
+                }
+                
+                if (!existingType.IsPrimitive)
+                {
+                    errors.EnsureError(DocumentErrorSeverity.Severe, arg, TexlStrings.ErrNeedPrimitive);
+                    isValid = false;
+                    continue;
+                }
+
+                var fError = false;
+                returnType = returnType.Add(ref fError, DPath.Root, columnName, existingType);
+                if (fError)
```

Comment:
>if (fError) [](http://example.com/codeflow?start=16&length=11)

is this where we'd catch it if we had a collision in names?
Sumarize(..., expr1 As Column1, expr2 As Column1)

Reply:
Yes. Test case has been added.